CreateProcessAsUser  Access is denied 0x00000005

timbrigham-oc opened the pull request:
Fix for "CreateProcessAsUser failed. (Access is denied, Win32ErrorCode 5 - 0x00000005)" bug. 

This is triggered if using Tanium for starting the tasks, at a minimum. 

References: 
https://social.technet.microsoft.com/Forums/en-US/89df4e34-4a70-4503-a4db-d8cee86e0c3b/createprocessasuser-failing-with-error-code-5-erroraccessdenied
https://learn.microsoft.com/en-us/windows/win32/procthread/process-creation-flags

timbrigham-oc committed:
Update runasuser.psm1

Fix for use in Tanium, corrects error "CreateProcessAsUser failed. (Access is denied, Win32ErrorCode 5 - 0x00000005)"

timbrigham-oc committed:
Merge pull request #1 from timbrigham-onecall/timbrigham-onecall-patch-1

Update runasuser.psm1

KelvinTegelaar commented:
This fix does make it get captured by EDR much faster in my tests, I'll see if we can improve this somehow.

timbrigham-oc commented:
Fair enough. I wasn't thinking too much about EDR being an issue. 
I utilize an internal repository and sign everything I want to publish, this included.

KelvinTegelaar commented:
Actually signing the module could be an improvement, I have a certificate so let me go test that. :)

jborean93 commented:
As an FYI this might fail if the current process is already in a job that doesn't allow breakway. It might be better to have this behind a switch to state that it should attempt to breakaway from the current job.

timbrigham-oc committed:
Merge branch 'KelvinTegelaar:master' into master

timbrigham-oc committed:
Merge branch 'KelvinTegelaar:master' into timbrigham-onecall-patch-1

timbrigham-oc committed:
Update runasuser.psm1

Add breakaway flag to StartProcessAsCurrentUser to set CREATE_BREAKAWAY_FROM_JOB

timbrigham-oc committed:
Update Invoke-AsCurrentUser.ps1

Pass breakaway flag to StartProcessAsCurrentUser

timbrigham-oc committed:
Update README.md

timbrigham-oc commented:
I agree on setting this up with a switch is a better route forward then just hardcoding it.  Based on "If the calling process is not associated with a job, this constant has no effect." it sounded harmless to set universally. 

It works perfectly in all my test cases, but changing the behavior and having it break on someone else isn't cool.  I can go ahead and wrap this up with a flag.

timbrigham-oc committed:
Update runasuser.psm1

Update statement for adding CREATE_BREAKAWAY_FROM_JOB, was missing alternate ref

timbrigham-oc committed:
Merge branch 'master' into timbrigham-onecall-patch-1

timbrigham-oc committed:
Merge pull request #2 from timbrigham-oc/timbrigham-onecall-patch-1

Timbrigham onecall patch 1

timbrigham-oc commented:
@jborean93, I wrapped it into a new flag "-Breakaway".. Does that look right?

jborean93 commented on runasuser.psm1:
```diff
@@ -370,7 +371,7 @@ namespace RunAsUser
             var startInfo = new NativeHelpers.STARTUPINFO();
             startInfo.cb = Marshal.SizeOf(startInfo);
-            uint dwCreationFlags = CREATE_UNICODE_ENVIRONMENT | (uint)(visible ? CREATE_NEW_CONSOLE : CREATE_NO_WINDOW);
+            uint dwCreationFlags = CREATE_UNICODE_ENVIRONMENT | (uint)(breakaway ? CREATE_BREAKAWAY_FROM_JOB : CREATE_UNICODE_ENVIRONMENT ) | (uint)(visible ? CREATE_NEW_CONSOLE : CREATE_NO_WINDOW);
```
can probably just use `breakaway ? CREATE_BREAKAWAY_FROM_JOB : 0` here to make it simple.

jborean93 reviewed (commented):
The code looks good, I'll probably mention in the docs that

> The `-Breakaway` switch will start the process with `CREATE_BREAKAWAY_FROM_JOB` which starts the process outside any [job](https://learn.microsoft.com/en-us/windows/win32/procthread/job-objects) the caller may be in.

timbrigham-oc committed:
Update README.md

timbrigham-oc committed:
Update runasuser.psm1

timbrigham-oc committed:
Merge pull request #3 from timbrigham-oc/timbrigham-onecall-patch-1

Timbrigham onecall patch 1

AC-JPB commented:
> Actually signing the module could be an improvement, I have a certificate so let me go test that. :)

Any word on getting the module signed? It's caused me a few headaches because it's unsigned.

baldsealion commented:
Stuff like opening OneDrive under the Invoke-RunAsUser command generate this error. Hope this fix gets included into the main module soon.

barren89 commented:
Hi,
When I run this with automate script then it is showing following error on some machines :-

invoke-ascurrentuser : Could not execute as currently logged on user: Exception calling "StartProcessAsCurrentUser" 
with "7" argument(s): "WTSQueryUserToken failed to get access token. (An attempt was made to reference a token that 
does not exist, Win32ErrorCode 1008 - 0x000003F0)"
At C:\windows\system32\config\systemprofile\AppData\Local\2a52a65ed1a44f4db0aa2d75b259738e.ps1:5 char:1
+ invoke-ascurrentuser -scriptblock $scriptblock -NoWait
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], MethodInvocationException
    + FullyQualifiedErrorId : System.Management.Automation.MethodInvocationException,Invoke-AsCurrentUser


Could you advise any solution for it?

Thanks,

AC-JPB commented:
@barren89 If you're using the "Execute script" option in Automate, it uses a custom powershell.exe and may have trouble installing modules, or running some commands. I've defaulted to using a shell command to execute the Windows powershell.exe with the -file switch to specify a ps1 script. IE 'C:\Windows\system32\WindowsPowerShell\v1.0\powershell.exe -ExecutionPolicy Bypass -File "<Path-To-Ps1>"

Just make sure the script has the installation of the RunAsUser module as part of it. Here's how I've done that in the past:
`If(!(Get-InstalledModule RunAsUser -ErrorAction silentlycontinue)){
  Install-Module -Name RunAsUser -Force -Confirm:$False}`

barren89 commented:
> @barren89 If you're using the "Execute script" option in Automate, it uses a custom powershell.exe and may have trouble installing modules, or running some commands. I've defaulted to using a shell command to execute the Windows powershell.exe with the -file switch to specify a ps1 script. IE `C:\Windows\system32\WindowsPowerShell\v1.0\powershell.exe -ExecutionPolicy Bypass -File "<Path-To-Ps1>"`
> 
> Just make sure the script has the installation of the RunAsUser module as part of it. Here's how I've done that in the past: `If(!(Get-InstalledModule RunAsUser -ErrorAction silentlycontinue)){ Install-Module -Name RunAsUser -Force -Confirm:$False}`

@AC-JPB I have verified, Module is installed successfully still it is giving that problem. I will try with shell command

KelvinTegelaar commented:
I thought I hit merge, not close. Thanks for tagging me.

KelvinTegelaar committed:
Merge branch 'master' into master

greatquux commented:
Hi - This is useful enough that it would be great if you could put out a new release on the PSGallery for automated install.  Thanks!

timbrigham-oc commented:
Good day all. I've been thinking about this for a while. :)
@greatquux , you especially, thanks for brining this to front of mind. 

Since it's been so long (and this is marked as MIT licensed!) I've forked a copy of the updates I made here - along with a few more recent pull requests - into RunAsUserBreakway in [PSGallery](https://www.powershellgallery.com/packages/RunAsUserBreakaway/2.4.0). If Kevin wants to get that module updated at some point I'll figure out how to get anyone installing my branched version back over to his version.
